transport/src/test/java/io/netty/channel/local/LocalChannelTest.java

```diff
@@ -390,6 +391,67 @@ public void operationComplete(ChannelFuture future) throws Exception {
         }
     }
 
+    @Test
+    public void testCloseAfterWriteInSameEventLoopPreservesOrder() throws InterruptedException {
+        Bootstrap cb = new Bootstrap();
+        ServerBootstrap sb = new ServerBootstrap();
+        final CountDownLatch messageLatch = new CountDownLatch(3);
+        final ByteBuf data = Unpooled.wrappedBuffer(new byte[1024]);
+
+        try {
+            cb.group(sharedGroup)
+                    .channel(LocalChannel.class)
+                    .handler(new ChannelInboundHandlerAdapter() {
+                        @Override
+                        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                            ctx.writeAndFlush(data.retainedDuplicate());
+                        }
+
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                            assertEquals(data, msg);
```

Comment:
consider surfacing this exception in the JUnit thread. One option is to set an AtomicReference<Throwable> and check it after the latch await returns.